schema.graphql

```diff
@@ -1,122 +1,124 @@
-  totalDeposited: BigDecimal!
-  totalWithdrawn: BigDecimal!
-  "totalActiveShares: totalSharesMinted - totalSharesBurned"
-  totalActiveShares: BigDecimal!
-  totalSharesMinted: BigDecimal!
-  totalSharesBurned: BigDecimal!
-  totalEarnings: BigDecimal!
-  # raw
-  pricePerFullShareRaw: BigInt!
-  totalSupplyRaw: BigInt!
-  # TODO:
-  "Balance of the Vault contract of underlying Token + balance of the Strategy contract of underlying Token"
-  vaultBalanceRaw: BigInt!
-  # TODO:
-  "Balance of underlying Token specifically held in the strategy"
-  strategyBalanceRaw: BigInt!
-  # TODO:
-  "How much the vault allows to be borrowed"
-  availableRaw: BigInt!
-  netDepositsRaw: BigInt!
-  totalDepositedRaw: BigInt!
-  totalWithdrawnRaw: BigInt!
-  totalActiveSharesRaw: BigInt!
-  totalSharesMintedRaw: BigInt!
-  totalSharesBurnedRaw: BigInt!
-  totalEarningsRaw: BigInt!
-  totalHarvestCalls: BigInt!
+
+  vaultBalanceUpdates: [VaultBalanceUpdate!]! @derivedFrom(field: "vault")
+
+  vaultUpdates: [VaultUpdate!]!
+  strategyUpdates: [StrategyUpdate!]!
+
   # derived fields
   transfers: [Transfer!]! @derivedFrom(field: "vault")
   deposits: [Deposit!]! @derivedFrom(field: "vault")
   withdrawals: [Withdrawal!]! @derivedFrom(field: "vault")
   harvests: [Harvest!]! @derivedFrom(field: "vault")
   balances: [AccountVaultBalance!]! @derivedFrom(field: "vault")
   strategies: [Strategy!]! @derivedFrom(field: "vault")
-  controllers: [Controller!]! @derivedFrom(field: "vault")
+  # controllers: [Controller!]! @derivedFrom(field: "vault")
+}
+
+# used to track balance of both user and vault
+interface BalanceUpdate @entity {
+  id: ID!
+  transaction: Transaction!
+
+  netDepositsRaw: BigInt!
+  totalDepositRaw: BigInt!
+  totalWithdrawalRaw: BigInt!
+  totalSharesMintedRaw: BigInt!
+  totalSharesBurnedRaw: BigInt!
+}
+
+type VaultBalanceUpdate implements BalanceUpdate @entity {
+  id: ID!
+  transaction: Transaction!
+
+  netDepositsRaw: BigInt!
+  totalDepositRaw: BigInt!
+  totalWithdrawalRaw: BigInt!
+  totalSharesMintedRaw: BigInt!
+  totalSharesBurnedRaw: BigInt!
+
+  vault: Vault!
+
+  totalActiveSharesRaw: BigInt!
+  totalEarningsRaw: BigInt!
+  totalHarvestCalls: BigInt!
+}
+
+type AccountBalanceUpdate implements BalanceUpdate @entity {
+  id: ID!
+  transaction: Transaction!
+
+  netDepositsRaw: BigInt!
+  totalDepositRaw: BigInt!
+  totalWithdrawalRaw: BigInt!
+  totalSharesMintedRaw: BigInt!
+  totalSharesBurnedRaw: BigInt!
+
+  account: Account!
+
+  totalSharesSentRaw: BigInt!
+  totalSharesReceivedRaw: BigInt!
+  totalSentRaw: BigInt!
+  totalReceivedRaw: BigInt!
+  shareBalanceRaw: BigInt!
+}
+
+type VaultUpdate @entity {
+  id: ID! # id + vault
+  transaction: Transaction!
+
+  vault: Vault!
+  pricePerFullShare: BigInt!
+  vaultBalance: BigInt!
+  totalSupply: BigInt!
+  available: BigInt!
+}
+
+type StrategyUpdate @entity {
+  id: ID! # id + strategy
+  transaction: Transaction!
+
+  strategy: Strategy!
+  strategyBalanceRaw: BigInt!
```

Comment:
StrategyUpdate probably does not need strategy balance. Balance can live on Strategy instead.